libs/schema-profunctor/README.md

```diff
@@ -203,6 +203,50 @@ Finally, we add a name to the schema using the `named`
 combinator. This does nothing to the JSON encoding-decoding part of
 the schema, and only affects the documentation.
 
+### Bespoke instances (Untagged sum type with overlapping values)
+
+Sometimes we have specific types handled in very specific ways, requiring
+a highly customised instance.
+
+Here's an example of such a type. We can have either Seconds or Unlimited,
+with a caveat: 0 ~ Unlimited.
+
+``` haskell
+data FeatureTTL
+  = FeatureTTLSeconds Word
+  | FeatureTTLUnlimited
+  deriving stock (Eq, Show)
+
+instance ToSchema FeatureTTL where
+  schema = mkSchema ttlDoc toTTL fromTTL
+    where
+      ttlDoc :: NamedSwaggerDoc
+      ttlDoc = swaggerDoc @Word & S.schema . S.example ?~ "unlimited"
+
+      toTTL :: A.Value -> A.Parser FeatureTTL
+      toTTL v = parseUnlimited v <|> parseSeconds v
+
+      parseUnlimited :: A.Value -> A.Parser FeatureTTL
+      parseUnlimited =
+        A.withText "FeatureTTL" $
+          \t ->
+            if t == "unlimited" || t == "0"
+              then pure FeatureTTLUnlimited
+              else A.parseFail "Expected ''unlimited' or '0'."
+
+      parseSeconds :: A.Value -> A.Parser FeatureTTL
+      parseSeconds = A.withScientific "FeatureTTL" $
+        \s -> case toBoundedInteger s of
+          Just 0 -> pure FeatureTTLUnlimited
+          Just i -> pure . FeatureTTLSeconds $ i
+          Nothing -> A.parseFail "Expected an integer."
+
+      fromTTL :: FeatureTTL -> Maybe A.Value
```

Comment:
Btw @pcapriotti why is this a `Maybe`?

Reply:
Good question. In order to support modular serialisation of sum types, serialisers can "fail" by returning `Nothing`. So for example, consider the serialiser side of how we build a schema for `A + B` from schemas `sa`, `sb` for `A` and `B` respectively. We first turn the serialiser of `sa` into a serialiser for `A + B` that simply fails in the `B` case, and similarly for `sb`. Then finally we merge them together using the appropriate `Monoid` instance for functions into `Maybe`. [This](https://github.com/wireapp/wire-server/blob/develop/libs/schema-profunctor/README.md#sum-types) has a bit more detail on the user-facing API of this mechanism.